File: day_10_kadanes_algorithm.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxSubarraySum(vector<int> &arr) {
        int n = arr.size();
        long long sum = 0, maxi = INT_MIN;
        
        for(int i = 0; i < n; i++) {
            sum += arr[i];
            if(sum > maxi) {
                maxi = max(sum, maxi);
            }
            if(sum < 0) {
                sum = 0;
            }
        }
        
        return maxi;
    }
};

int main() {
    int n;
    cin >> n;
    vector<int> arr;

    for(int i = 0; i < n; i++) {
        int x;
        cin >> x;
        arr.push_back(x);
    }

    Solution sol;
    cout << sol.maxSubarraySum(arr) << endl;

    return 0;
}

File: Day10KadanesAlgorithm.java
import java.util.Scanner;

public class Day10KadanesAlgorithm {
    private class Solution {
        int maxSubarraySum(int[] arr) {
            int n = arr.length;
            int sum = 0, maxi = Integer.MIN_VALUE;
            
            for(int i = 0; i < n; i++) {
                sum += arr[i];
                if(sum > maxi) {
                    maxi = sum;
                }
                if(sum < 0) {
                    sum = 0;
                }
            }
            
            return maxi;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] arr = new int[n];

        for(int i = 0; i < n; i++) {
            arr[i] = scanner.nextInt();
        }

        Solution sol;
        sol = new Day10KadanesAlgorithm().new Solution();
        System.out.println(sol.maxSubarraySum(arr));

        scanner.close();
    }
}
